### PR TITLE
BUG: Fix for SVD not always sorted with hermitian=True

### LINKED ISSUE #15462: np.linalg.svd with hermitian=True does not properly sort singular values
### Reproducing code example:

`svd` with `hermitian=True` (introduced by @eric-wieser in https://github.com/numpy/numpy/pull/12693) has a bug with the order of singular values:
```python
import numpy as np

a = np.array([[0, 1, 0], [1, 0, 1], [0, 1, 0]])
u, s, v = np.linalg.svd(a, hermitian=True)
print(s)
```
Outputs: `[1.41421356e+00 6.07153217e-18 1.41421356e+00]`

Per the contract of `svd`, the singular values should be sorted in descending order, but they aren't here. That's because we simply reverse the order of eigenvalues than sorting, which only suffices if all eigenvalues happen to be positive.

### Numpy/Python version information:

1.17.5 3.6.9 (default, Nov  7 2019, 10:44:02) 
[GCC 8.3.0]

### PR DESCRIPTION
**Pull request by hameerabbasi**
Fixes #15462

### CODE REVIEW

File: numpy/linalg/linalg.py [hunk 1]
```diff
@@ -1615,17 +1615,20 @@ def svd(a, full_matrices=True, compute_uv=True, hermitian=False):
         # reversing is cheap by design in numpy, so we do so to be consistent
         if compute_uv:
             s, u = eigh(a)
-            s = s[..., ::-1]
-            u = u[..., ::-1]
-            # singular values are unsigned, move the sign into v
-            vt = transpose(u * sign(s)[..., None, :]).conjugate()
+            sgn = sign(s)
             s = abs(s)
+            sidx = s.shape[-1] - 1 - argsort(s)
```
Comment (eric-wieser) [inline numpy/linalg/linalg.py]: Doesn't looks correct to me. Shouldn't this be the following?
```suggestion
            sidx = argsort(s)[::-1]
```

File: numpy/linalg/tests/test_linalg.py [hunk 2]
```diff
@@ -680,6 +680,7 @@ def do(self, a, b, tags):
         assert_allclose(a, dot_generalized(np.asarray(u) * np.asarray(s)[..., None, :],
                                            np.asarray(vt)),
                         rtol=get_rtol(u.dtype))
+        assert_equal(np.sort(s)[::-1], s)
```
Comment (eric-wieser) [inline numpy/linalg/tests/test_linalg.py]: This looks unlikely to work for multi-dimensional `s`

File: numpy/linalg/linalg.py (hunk 1 above)
Comment (hameerabbasi) [reply numpy/linalg/linalg.py]: Yes, that is what I was thinking of.

File: numpy/linalg/linalg.py [hunk 3]
```diff
@@ -1615,17 +1615,20 @@ def svd(a, full_matrices=True, compute_uv=True, hermitian=False):
         # reversing is cheap by design in numpy, so we do so to be consistent
         if compute_uv:
             s, u = eigh(a)
-            s = s[..., ::-1]
-            u = u[..., ::-1]
-            # singular values are unsigned, move the sign into v
-            vt = transpose(u * sign(s)[..., None, :]).conjugate()
+            sgn = sign(s)
             s = abs(s)
+            sidx = s.shape[-1] - 1 - argsort(s[::-1])
```
Comment (eric-wieser) [inline numpy/linalg/linalg.py]: Needs a `...`

While correct ignoring that, this is slower than just `argsort(s)[..., ::-1]`, and we really don't care about how ties are resolved.

File: numpy/linalg/linalg.py [hunk 4]
```diff
@@ -1615,17 +1615,20 @@ def svd(a, full_matrices=True, compute_uv=True, hermitian=False):
...
-            s = s[..., ::-1]
-            u = u[..., ::-1]
-            # singular values are unsigned, move the sign into v
-            vt = transpose(u * sign(s)[..., None, :]).conjugate()
+            sgn = sign(s)
             s = abs(s)
+            sidx = argsort(s)[..., ::-1]
+            sgn = sgn[..., sidx]
+            s = s[..., sidx]
+            u = u[..., sidx]
```
Comment (eric-wieser) [inline numpy/linalg/linalg.py]: I might be wrong, but do these need to use `take_along_axis`?

File: numpy/linalg/linalg.py [hunk 5]
```diff
@@ -1615,17 +1615,20 @@ def svd(a, full_matrices=True, compute_uv=True, hermitian=False):
         # reversing is cheap by design in numpy, so we do so to be consistent
```
Comment (eric-wieser) [inline numpy/linalg/linalg.py]: This comment needs updating

File: numpy/linalg/linalg.py (hunk 4 above)
Comment (hameerabbasi) [reply numpy/linalg/linalg.py]: Axis is always `-1` in the sort, so I really don't think it matters.

Comment (eric-wieser) [reply numpy/linalg/linalg.py]:
```python
>>> a = np.arange(12).reshape(3, 4)
>>> a[..., a.argsort()].shape
(3, 3, 4)
>>> np.take_along_axis(a, a.argsort(), axis=-1).shape
(3, 4)
```

It does matter.

Comment (hameerabbasi) [reply numpy/linalg/linalg.py]: Looks like `np.lib.index_base` is half-initialized when `linalg` is initialized, so I have to do the import inside the function.

File: numpy/linalg/linalg.py [hunk 6]
```diff
@@ -1608,24 +1608,29 @@ def svd(a, full_matrices=True, compute_uv=True, hermitian=False):
     True
 
     """
+    import numpy as _nx
     a, wrap = _makearray(a)
 
     if hermitian:
-        # note: lapack returns eigenvalues in reverse order to our contract.
-        # reversing is cheap by design in numpy, so we do so to be consistent
+        # note: lapack returns eigenvalues with s ** 2 sorted instead of s.
```
Comment (eric-wieser) [inline numpy/linalg/linalg.py]:
```suggestion
        # note: lapack svd returns eigenvalues with s ** 2 sorted descending, but eig
        # returns s sorted ascending.
```
or similar

File: numpy/linalg/linalg.py [hunk 7]
```diff
@@ -1608,24 +1608,29 @@ def svd(a, full_matrices=True, compute_uv=True, hermitian=False):
...
-            s = s[..., ::-1]
-            u = u[..., ::-1]
-            # singular values are unsigned, move the sign into v
-            vt = transpose(u * sign(s)[..., None, :]).conjugate()
+            sgn = sign(s)
             s = abs(s)
+            sidx = argsort(s)[..., ::-1]
+            sgn = _nx.take_along_axis(sgn, sidx, axis=-1)
+            s = _nx.take_along_axis(s, sidx, axis=-1)
+            u = _nx.take_along_axis(u, sidx[..., None, :], axis=-1)
```
Comment (eric-wieser) [inline numpy/linalg/linalg.py]: I'm not convinced whether this is the right way around (`None, :` vs `:, None`). 

Do tests fail if you swap them?

Comment (hameerabbasi) [reply numpy/linalg/linalg.py]: > Do tests fail if you swap them?

I just checked, and they do. Reason about it this way: The last dimension (columns) is the one that has to be sorted, so we have to place the last axis of `sidx` parallel to it.

This is matching with the `None, :` below. If you like I could add a comment.

**Review by eric-wieser — APPROVED** [review]
Looks pretty reasonable. @shoyer, mind taking a look too?

**Comment by shoyer** [comment]
Could we add a test that verified that we get the same SVD result (up to floating point error) if we pass a random hermitian matrix into SVD either with or without hermitian=True?

**Comment by hameerabbasi** [comment]
> Could we add a test that verified that we get the same SVD result (up to floating point error) if we pass a random hermitian matrix into SVD either with or without hermitian=True?

I believe such a test would be incorrect since the SVD isn’t unique. See https://math.stackexchange.com/a/3362413

**Comment by shoyer** [comment]
> > Could we add a test that verified that we get the same SVD result (up to floating point error) if we pass a random hermitian matrix into SVD either with or without hermitian=True?
> 
> I believe such a test would be incorrect since the SVD isn’t unique. See https://math.stackexchange.com/a/3362413

Yes, but at least we could check that the singular values are identical. Could we at least check those, for `compute_uv=True` and `compute_uv=False`?

**Comment by seberg** [comment]
It is not generally unique, although I would imagine if you use e.g. `multivariate_random` with well enough defined data, it is probably also possible to achieve an "almost" unique result?

**Comment by shoyer** [comment]
If the singular values are unique, the `u` and `v` matrices will also be unique up to global phases. So you could safely compare `abs(u)`.

**Comment by hameerabbasi** [comment]
Instead of testing some arbitrary input from some arbitrary function, I've ensured that:

1. The singular values are sorted.
2. That `u` and `vt` are Hermitian.

In addition, the function already tested that the product produced was correct. These things, when combined, test the definition of the SVD and the order of the singular values guaranteed in the documentation.

**Comment by hameerabbasi** [comment]
I think the Azure CI failure is a fluke, given that the test isn't related to SVD but rather to matrix multiplication. Does someone have the permissions to run that without running the whole test suite, or can this be merged without it?

**Comment by hameerabbasi** [comment]
CI is green, is anything else needed here?

**Comment by mattip** [comment]
Thanks @hameerabbasi

**Pull request merged**